src/hooks/useContractAbiItems.ts

```diff
@@ -0,0 +1,75 @@
+import { type Abi } from "abitype";
+import { thirdwebClient } from "lib/thirdweb-client";
+import { useEffect, useState } from "react";
+import { Address, defineChain, getContract } from "thirdweb";
+import { resolveContractAbi } from "thirdweb/contract";
+
+/**
+ * Returns read/write function and event names from a contract.
+ *
+ * @param chainId number
+ * @param address Address
+ * @returns readFunctions - List of read function names
+ * @returns writeFunctinos - List of write function names
+ * @returns events - List of event names
+ */
+export const useContractAbiItems = (
+  chainId: number,
+  address: Address,
+): {
+  readFunctions: string[] | undefined;
+  writeFunctions: string[] | undefined;
+  events: string[] | undefined;
+} => {
+  const [readFunctions, setReadFunctions] = useState<string[] | undefined>();
+  const [writeFunctions, setWriteFunctions] = useState<string[] | undefined>();
+  const [events, setEvents] = useState<string[] | undefined>();
+
+  useEffect(() => {
+    const chain = defineChain(chainId);
+    const contract = getContract({
+      client: thirdwebClient,
+      chain,
+      address,
+    });
+
+    const dedupeAndSort = (arr: string[]) => [...new Set(arr)].toSorted();
+
+    (async () => {
+      const _readFunctions: string[] = [];
+      const _writeFunctions: string[] = [];
+      const _events: string[] = [];
+
+      try {
+        const abi = await resolveContractAbi<Abi>(contract);
+        for (const abiItem of abi) {
+          if (abiItem.type === "function") {
+            if (
+              abiItem.stateMutability === "pure" ||
+              abiItem.stateMutability === "view"
+            ) {
+              _readFunctions.push(abiItem.name);
+            } else {
+              _writeFunctions.push(abiItem.name);
+            }
+          } else if (abiItem.type === "event") {
+            _events.push(abiItem.name);
+          }
+        }
+
+        // Dedupe and set the hook state.
+        setReadFunctions(dedupeAndSort(_readFunctions));
+        setWriteFunctions(dedupeAndSort(_writeFunctions));
+        setEvents(dedupeAndSort(_events));
+      } catch (e) {
+        console.error(
+          `Failed to load ABI for contract ${address} on chain ${chainId}.`,
+        );
+        setWriteFunctions([]);
+        setEvents([]);
+      }
+    })();
+  }, [chainId, address]);
+
+  return { readFunctions, writeFunctions, events };
```

Comment:
this entire thing should be a `useQuery` instead: `chainId` and `contractAddress` should be part of the query key

something along these lines (rough):
```
const {data, isLoading} = useQuery({
  queryKey: ['contract-abi-items', chainId, address],
  queryFn: async () => {
// ... your async logic here, return the readFunctions, writeFunctions, events from it at the end
})

return data;
```